aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,329 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
+  if (matrices.dim() > 2) {
+    auto res = at::empty(matrices.sizes(), matrices.options());
+
+    batch_apply(
+      {res, matrices, powers},
+      matrices.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+        int64_t n = tensors[2].item().template to<int64_t>();
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + A2 / 2;
+}
+
+Tensor compute_T4(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / 2. + A / 6. + A2 / 24.);
+}
+
+template <typename scalar_t>
+Tensor compute_T8(const Tensor& A) {
+  // no constexpr support for std::sqrt unfortunately
+  scalar_t sqrt_177 = std::sqrt(177);
+  scalar_t x3 = 2. / 3.;
+  scalar_t x1 = x3 * ((1 + sqrt_177) / 88);
+  scalar_t x2 = x3 * ((1 + sqrt_177) / 352);
+  scalar_t x4 = (-271 + 29 * sqrt_177) / (315 * x3);
+  scalar_t x5 = (-11 + 11 * sqrt_177) / (1260 * x3);
+  scalar_t x6 = (-99 + 11 * sqrt_177) / (5040 * x3);
+  scalar_t x7 = (89 - sqrt_177) / (5040 * x3);
+  scalar_t y2 = (857 - 58 * sqrt_177) / 630;
+
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  auto A4 = at::matmul(A2, x1 * A + x2 * A2);
+  auto A8 = at::matmul(x3 * A2 + A4, x4 * I + x5 * A + x6 * A2 + x7 * A4);
+  return I + A + y2 * A2 + A8;
+}
+
+template <typename scalar_t>
+Tensor compute_T12(const Tensor& A) {
+  constexpr scalar_t b[][4] = {
+    {
+      9.0198e-16,
+      0.46932117595418237389,
+      -0.20099424927047284052,
+      -0.04623946134063071740
+    },
+    {
+      5.31597895759871264183,
+      1.19926790417132231573,
+      0.01179296240992997031,
+      0.01108844528519167989
+    },
+    {
+      0.18188869982170434744,
+      0.05502798439925399070,
+      0.09351590770535414968,
+      0.00610700528898058230
+    },
+    {
+      -2.0861320e-13,
+      -0.13181061013830184015,
+      -0.02027855540589259079,
+      -0.00675951846863086359
+    }
+  };
+
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  auto A3 = at::matmul(A2, A);
+  std::reference_wrapper<const Tensor> As[] = {I, A, A2, A3};
+
+  Tensor Bs[4];
+  for (int i = 0; i < 4; ++i) {
+    Bs[i] = at::zeros(A.sizes(), A.options());
+  }
+
+  for (int i = 0; i < 4; ++i) {
```

Comment:
might be worth  doing this as a matrix-matrix pointwise multiplication?

Reply:
I am not sure this is efficient. In your solution you stack the matrices, and I have doubts that the storage will be shared.

Reply:
Fair enough. Now that you mention that, I am not sure what could be best. Maybe someone with more AT experience could comment on this?

Reply:
The best is to do this loop on a specific device, which I plan on doing.